{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## SNAP proxy search for SNPs in LD with lead SNPs\n",
    "\n",
    "1. Navigate to [SNAP Proxy Search](https://www.broadinstitute.org/mpg/snap/ldsearch.php)\n",
    "+ Enter lead SNPs of interest under \"Query SNPs\"\n",
    "+ Under \"Search Options\" set\n",
    "  + SNP data set: 1000 Genomes Pilot 1\n",
    "  + Population panel: CEU\n",
    "  + r2 threshold: 0.5\n",
    "  + Distance limit: 500\n",
    "+ Under \"Ouput Options\"\n",
    "  + Download to: File\n",
    "  + Select \"Include each query snp as a proxy for itself\"\n",
    "  + Select \"Suppress warning messages in output\"\n",
    "+ Under \"Filter By Array\" make sure all are unselected\n",
    "+ Under \"Output Columns\" select all besides \"Associated Gene Annotations from GeneCruiser (decreases performance)\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "import pandas\n",
    "import numpy\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn\n",
    "\n",
    "%matplotlib inline"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div style=\"max-height:1000px;max-width:1500px;overflow:auto;\">\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SNP</th>\n",
       "      <th>Proxy</th>\n",
       "      <th>Distance</th>\n",
       "      <th>RSquared</th>\n",
       "      <th>DPrime</th>\n",
       "      <th>Arrays</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Coordinate_HG18</th>\n",
       "      <th>RecombinationRate</th>\n",
       "      <th>GeneticMapDistance</th>\n",
       "      <th>GeneticMapPosition</th>\n",
       "      <th>Major</th>\n",
       "      <th>Minor</th>\n",
       "      <th>MAF</th>\n",
       "      <th>NObserved</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>rs1198588</td>\n",
       "      <td>rs1198588</td>\n",
       "      <td>0</td>\n",
       "      <td>1.000</td>\n",
       "      <td>1</td>\n",
       "      <td>AN,A5,A6,ICA,ICB</td>\n",
       "      <td>chr1</td>\n",
       "      <td>98325420</td>\n",
       "      <td>0.27184</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>131.11046</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>0.242</td>\n",
       "      <td>120</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>rs1198588</td>\n",
       "      <td>rs1782812</td>\n",
       "      <td>5330</td>\n",
       "      <td>0.955</td>\n",
       "      <td>1</td>\n",
       "      <td>AN,A5,A6,I3,I5,I6,I6Q,IM,IMD,IC,ICQ,IWQ,O54,O5...</td>\n",
       "      <td>chr1</td>\n",
       "      <td>98320090</td>\n",
       "      <td>0.26964</td>\n",
       "      <td>0.00157</td>\n",
       "      <td>131.10889</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.233</td>\n",
       "      <td>120</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>rs1198588</td>\n",
       "      <td>rs1198581</td>\n",
       "      <td>22273</td>\n",
       "      <td>0.911</td>\n",
       "      <td>1</td>\n",
       "      <td>CYT,OQ,OE,O24,O28,O54,O5E,OEE</td>\n",
       "      <td>chr1</td>\n",
       "      <td>98303147</td>\n",
       "      <td>0.30153</td>\n",
       "      <td>0.00611</td>\n",
       "      <td>131.10435</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>0.225</td>\n",
       "      <td>120</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>rs1198588</td>\n",
       "      <td>rs2660299</td>\n",
       "      <td>24380</td>\n",
       "      <td>0.911</td>\n",
       "      <td>1</td>\n",
       "      <td>A6,OQ,OE,O24,O28,O54,O5E,OEE,ICA,ICB</td>\n",
       "      <td>chr1</td>\n",
       "      <td>98301040</td>\n",
       "      <td>0.16161</td>\n",
       "      <td>0.00648</td>\n",
       "      <td>131.10398</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>0.225</td>\n",
       "      <td>120</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>rs1198588</td>\n",
       "      <td>rs2660300</td>\n",
       "      <td>24621</td>\n",
       "      <td>0.911</td>\n",
       "      <td>1</td>\n",
       "      <td>I3,I5,I6,I6Q,IM,IMD,IC,ICQ,IWQ,O54,O5E,ICA,ICB</td>\n",
       "      <td>chr1</td>\n",
       "      <td>98300799</td>\n",
       "      <td>0.15214</td>\n",
       "      <td>0.00652</td>\n",
       "      <td>131.10394</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>0.225</td>\n",
       "      <td>120</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         SNP      Proxy  Distance  RSquared  DPrime  \\\n",
       "0  rs1198588  rs1198588         0     1.000       1   \n",
       "1  rs1198588  rs1782812      5330     0.955       1   \n",
       "2  rs1198588  rs1198581     22273     0.911       1   \n",
       "3  rs1198588  rs2660299     24380     0.911       1   \n",
       "4  rs1198588  rs2660300     24621     0.911       1   \n",
       "\n",
       "                                              Arrays Chromosome  \\\n",
       "0                                   AN,A5,A6,ICA,ICB       chr1   \n",
       "1  AN,A5,A6,I3,I5,I6,I6Q,IM,IMD,IC,ICQ,IWQ,O54,O5...       chr1   \n",
       "2                      CYT,OQ,OE,O24,O28,O54,O5E,OEE       chr1   \n",
       "3               A6,OQ,OE,O24,O28,O54,O5E,OEE,ICA,ICB       chr1   \n",
       "4     I3,I5,I6,I6Q,IM,IMD,IC,ICQ,IWQ,O54,O5E,ICA,ICB       chr1   \n",
       "\n",
       "   Coordinate_HG18  RecombinationRate  GeneticMapDistance  GeneticMapPosition  \\\n",
       "0         98325420            0.27184             0.00000           131.11046   \n",
       "1         98320090            0.26964             0.00157           131.10889   \n",
       "2         98303147            0.30153             0.00611           131.10435   \n",
       "3         98301040            0.16161             0.00648           131.10398   \n",
       "4         98300799            0.15214             0.00652           131.10394   \n",
       "\n",
       "  Major Minor    MAF  NObserved  \n",
       "0     T     A  0.242        120  \n",
       "1     G     A  0.233        120  \n",
       "2     C     T  0.225        120  \n",
       "3     T     C  0.225        120  \n",
       "4     C     A  0.225        120  "
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Read SNAP proxy search\n",
    "path = 'data/snap/do-slim-proxy-SNPs.tsv.gz'\n",
    "snap_df = pandas.read_table(path, compression='gzip', na_values=['N/A'])\n",
    "snap_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "38"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Condense SNAP proxy search by lead SNP and compute windows\n",
    "snap_errors = list()\n",
    "\n",
    "rows = list()\n",
    "for snp, group_df in snap_df.groupby('SNP', as_index=False):\n",
    "    try:\n",
    "        lead = group_df[group_df.Proxy == snp].iloc[0]\n",
    "    except IndexError:\n",
    "        snap_errors.append(snp)\n",
    "        continue\n",
    "    \n",
    "    row = pandas.Series()\n",
    "    row['lead_snp'] = snp\n",
    "    row['lead_chrom'] = lead['Chromosome']\n",
    "    row['lead_coord'] = lead['Coordinate_HG18']\n",
    "    row['lead_maf'] = lead['MAF']\n",
    "    \n",
    "    window_df = group_df[(group_df.Chromosome == row['lead_chrom']) & (group_df.RSquared >= 0.8)]\n",
    "    i_min = numpy.argmin(list(window_df.Coordinate_HG18))\n",
    "    i_max = numpy.argmax(list(window_df.Coordinate_HG18))\n",
    "    for side, i in ('lower', i_min), ('upper', i_max):\n",
    "        window = window_df.iloc[i]\n",
    "        row[side + '_snp'] = window.Proxy\n",
    "        row[side + '_coord'] = window.Coordinate_HG18\n",
    "        row[side + '_map_dist'] = window.GeneticMapDistance\n",
    "        row[side + '_maf'] = window.MAF\n",
    "\n",
    "    rows.append(row)\n",
    "\n",
    "window_df = pandas.DataFrame(rows)\n",
    "len(snap_errors)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div style=\"max-height:1000px;max-width:1500px;overflow:auto;\">\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>lead_snp</th>\n",
       "      <th>lead_chrom</th>\n",
       "      <th>lead_coord</th>\n",
       "      <th>lead_maf</th>\n",
       "      <th>lower_snp</th>\n",
       "      <th>lower_coord</th>\n",
       "      <th>lower_map_dist</th>\n",
       "      <th>lower_maf</th>\n",
       "      <th>upper_snp</th>\n",
       "      <th>upper_coord</th>\n",
       "      <th>upper_map_dist</th>\n",
       "      <th>upper_maf</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>rs1000113</td>\n",
       "      <td>chr5</td>\n",
       "      <td>150220269</td>\n",
       "      <td>0.042</td>\n",
       "      <td>rs76767593</td>\n",
       "      <td>150150036</td>\n",
       "      <td>0.04616</td>\n",
       "      <td>0.042</td>\n",
       "      <td>rs73268545</td>\n",
       "      <td>150315475</td>\n",
       "      <td>0.02666</td>\n",
       "      <td>0.042</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>rs1000579</td>\n",
       "      <td>chr4</td>\n",
       "      <td>4770395</td>\n",
       "      <td>0.442</td>\n",
       "      <td>rs1000579</td>\n",
       "      <td>4770395</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.442</td>\n",
       "      <td>rs10012526</td>\n",
       "      <td>4778062</td>\n",
       "      <td>0.00758</td>\n",
       "      <td>0.425</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>rs1000589</td>\n",
       "      <td>chr13</td>\n",
       "      <td>63039914</td>\n",
       "      <td>0.450</td>\n",
       "      <td>rs672064</td>\n",
       "      <td>62935367</td>\n",
       "      <td>0.05226</td>\n",
       "      <td>0.483</td>\n",
       "      <td>rs7323601</td>\n",
       "      <td>63042754</td>\n",
       "      <td>0.00090</td>\n",
       "      <td>0.450</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>rs1000597</td>\n",
       "      <td>chr7</td>\n",
       "      <td>30903703</td>\n",
       "      <td>0.100</td>\n",
       "      <td>rs10229128</td>\n",
       "      <td>30899051</td>\n",
       "      <td>0.01045</td>\n",
       "      <td>0.100</td>\n",
       "      <td>rs1000597</td>\n",
       "      <td>30903703</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.100</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>rs10007052</td>\n",
       "      <td>chr4</td>\n",
       "      <td>142225023</td>\n",
       "      <td>0.183</td>\n",
       "      <td>rs28529555</td>\n",
       "      <td>142224374</td>\n",
       "      <td>0.00013</td>\n",
       "      <td>0.183</td>\n",
       "      <td>rs11734545</td>\n",
       "      <td>142231391</td>\n",
       "      <td>0.00587</td>\n",
       "      <td>0.175</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     lead_snp lead_chrom  lead_coord  lead_maf   lower_snp  lower_coord  \\\n",
       "0   rs1000113       chr5   150220269     0.042  rs76767593    150150036   \n",
       "1   rs1000579       chr4     4770395     0.442   rs1000579      4770395   \n",
       "2   rs1000589      chr13    63039914     0.450    rs672064     62935367   \n",
       "3   rs1000597       chr7    30903703     0.100  rs10229128     30899051   \n",
       "4  rs10007052       chr4   142225023     0.183  rs28529555    142224374   \n",
       "\n",
       "   lower_map_dist  lower_maf   upper_snp  upper_coord  upper_map_dist  \\\n",
       "0         0.04616      0.042  rs73268545    150315475         0.02666   \n",
       "1         0.00000      0.442  rs10012526      4778062         0.00758   \n",
       "2         0.05226      0.483   rs7323601     63042754         0.00090   \n",
       "3         0.01045      0.100   rs1000597     30903703         0.00000   \n",
       "4         0.00013      0.183  rs11734545    142231391         0.00587   \n",
       "\n",
       "   upper_maf  \n",
       "0      0.042  \n",
       "1      0.425  \n",
       "2      0.450  \n",
       "3      0.100  \n",
       "4      0.175  "
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "window_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "ename": "AttributeError",
     "evalue": "'DataFrame' object has no attribute 'maf'",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mAttributeError\u001b[0m                            Traceback (most recent call last)",
      "\u001b[1;32m<ipython-input-5-9902baa4959e>\u001b[0m in \u001b[0;36m<module>\u001b[1;34m()\u001b[0m\n\u001b[0;32m      1\u001b[0m \u001b[1;31m# View window span (kilobases) versus minor allele frequency\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m----> 2\u001b[1;33m \u001b[0mseaborn\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mjointplot\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mwindow_df\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mmaf\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mwindow_df\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mspan_kb\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mkind\u001b[0m\u001b[1;33m=\u001b[0m\u001b[1;34m'hex'\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m;\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[1;32m/home/dhimmels/anaconda3/envs/my-rdkit-env/lib/python3.4/site-packages/pandas/core/generic.py\u001b[0m in \u001b[0;36m__getattr__\u001b[1;34m(self, name)\u001b[0m\n\u001b[0;32m   2081\u001b[0m                 \u001b[1;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[1;33m[\u001b[0m\u001b[0mname\u001b[0m\u001b[1;33m]\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m   2082\u001b[0m             raise AttributeError(\"'%s' object has no attribute '%s'\" %\n\u001b[1;32m-> 2083\u001b[1;33m                                  (type(self).__name__, name))\n\u001b[0m\u001b[0;32m   2084\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m   2085\u001b[0m     \u001b[1;32mdef\u001b[0m \u001b[0m__setattr__\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mself\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mname\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mvalue\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;31mAttributeError\u001b[0m: 'DataFrame' object has no attribute 'maf'"
     ]
    }
   ],
   "source": [
    "# View window span (kilobases) versus minor allele frequency\n",
    "seaborn.jointplot(window_df.maf, window_df.span_kb, kind='hex');"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.4.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
